app/src/test/kotlin/com/hedvig/app/feature/genericauth/OtpInputViewModelTest.kt

```diff
@@ -1,192 +1,159 @@
+      viewModel.submitCode("123456")
+      runCurrent()
+      assertThat(viewModel.viewState.value.loadingCode).isEqualTo(true)
+      assertThat(viewModel.viewState.value.loadingResend).isEqualTo(false)
+
+      authRepository.submitOtpResponse.add(SubmitOtpResult.Success(AuthorizationCodeGrant("")))
+      authRepository.exchangeResponse.add(AuthTokenResult.Success(accessToken, refreshToken))
+      runCurrent()
+
+      assertThat(viewModel.viewState.value.loadingCode).isEqualTo(false)
+      authTokenStorage.getTokens().first().also { (resultAccessToken, resultRefreshToken) ->
+        assertThat(resultAccessToken?.token).isEqualTo(accessToken.token)
+        assertThat(resultRefreshToken?.token).isEqualTo(refreshToken.token)
+      }
+      val event = awaitItem()
+      assertThat(event).isEqualTo(OtpInputViewModel.Event.Success(accessToken.token))
     }
-
-    viewModel.submitCode("123456")
-    assertThat(viewModel.viewState.value.loadingCode).isEqualTo(true)
-    assertThat(viewModel.viewState.value.loadingResend).isEqualTo(false)
-
-    advanceUntilIdle()
-    assertThat(viewModel.viewState.value.loadingCode).isEqualTo(false)
-    assertThat(authToken).isEqualTo("testToken")
-    assertThat(events.size).isEqualTo(1)
-    assertThat(events.first()).isEqualTo(OtpInputViewModel.Event.Success("testToken"))
-    eventCollectingJob.cancel()
   }
 
   @Test
   fun testResendError() = runTest {
-    resendOtpResult = com.hedvig.android.auth.ResendOtpResult.Error("Error")
+    val authRepository = FakeAuthRepository()
+    val viewModel = testViewModel(authRepository)
+    val errorMessage = "Error"
 
     viewModel.resendCode()
+    runCurrent()
     assertThat(viewModel.viewState.value.loadingCode).isEqualTo(false)
     assertThat(viewModel.viewState.value.loadingResend).isEqualTo(true)
 
-    advanceUntilIdle()
-    assertThat(viewModel.viewState.value.networkErrorMessage).isEqualTo("Error")
+    authRepository.resendOtpResponse.add(ResendOtpResult.Error(errorMessage))
+    runCurrent()
+    assertThat(viewModel.viewState.value.networkErrorMessage).isEqualTo(errorMessage)
   }
 
   @Test
   fun testResend() = runTest {
-    resendOtpResult = com.hedvig.android.auth.ResendOtpResult.Success
-
-    val events = mutableListOf<OtpInputViewModel.Event>()
-    val eventCollectingJob = launch {
-      viewModel.events.collect { events.add(it) }
+    val authRepository = FakeAuthRepository()
+    val viewModel = testViewModel(authRepository)
+
+    viewModel.events.test {
+      viewModel.resendCode()
+      runCurrent()
+      assertThat(viewModel.viewState.value.loadingCode).isEqualTo(false)
+      assertThat(viewModel.viewState.value.loadingResend).isEqualTo(true)
+
+      authRepository.resendOtpResponse.add(ResendOtpResult.Success)
+      runCurrent()
+      assertThat(viewModel.viewState.value.loadingResend).isEqualTo(false)
+      val event = awaitItem()
+      assertThat(event).isEqualTo(OtpInputViewModel.Event.CodeResent)
     }
```

Comment:
So the reason I changed these a tiny bit is as follows:

If we call `runCurrent()` directly after `resetCode()` is called, we'd get the wrong assertions for 
```
assertThat(viewModel.viewState.value.loadingCode).isEqualTo(false)
assertThat(viewModel.viewState.value.loadingResend).isEqualTo(true)
```
since they would not be true if the code contained in the `viewmodelScope.launch` inside `resendCode()` would run before them.
So in the test previously we depended on the way the test schedules the jobs for this assertion to work, basically making the test racy, meaning if we changed the impl details of how the coroutines were launched (for example if we moved the .update of the viewState inside the viewModelScope.launch{} block this test would fail) we might get different results.

This change makes it so that after we call `resendCode()` we call `runCurrent()` to let it do everything it's supposed to do, and then we assert that we're in the loading state since we know that authRepository has *not* responded anything yet, and we know that since we haven't fed anything to the `authRepository.resendOtpResponse` yet. *Then* we give it the response, `runCurrent()` to let OtpInputViewModel to process whatever that means for it when the repository has responded with something, and then we can safely make our assertions again. Nice and one by one, making reasoning about what's happening in which order a bit simpler rather than relying on how the test coroutine scheduling might work or not.

For example if in the previous test we changed the dispatcher it was running on to an `UnconfinedTestDispatcher()` it'd start failing since we wouldn't be able to see this in-between loading state, with this approach this would still pass since we're also controlling when the authRepository responds with something.

Reply:
Makes sense 👍